core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,692 @@
+                assert_eq!(
+                    *lock_thread_id, thread_id,
+                    "outstanding write lock must be on same thread"
+                );
+                *lock_count -= 1;
+                if *lock_count == 0 {
+                    entry.remove();
+                }
+            }
+            Entry::Vacant(_) => {
+                panic!("write lock must exist for account: {account}");
+            }
+        }
+    }
+
+    /// Locks the given `account` for reading on `thread_id`.
+    /// Panics if the account is already locked for writing on another thread.
+    fn read_lock_account(&mut self, account: &Pubkey, thread_id: ThreadId) {
+        match self.read_locks.entry(*account) {
+            Entry::Occupied(mut entry) => {
+                let (thread_set, lock_counts) = entry.get_mut();
+                thread_set.insert(thread_id);
+                lock_counts[thread_id] += 1;
+            }
+            Entry::Vacant(entry) => {
+                let mut lock_counts = [0; MAX_THREADS];
+                lock_counts[thread_id] = 1;
+                entry.insert((ThreadSet::only(thread_id), lock_counts));
+            }
+        }
+
+        // Check for outstanding write-locks
+        if let Some((write_thread_id, _)) = self.write_locks.get(account) {
+            assert_eq!(
+                write_thread_id, &thread_id,
+                "outstanding write lock must be on same thread"
+            );
+        }
+    }
+
+    /// Unlocks the given `account` for reading on `thread_id`.
+    /// Panics if the account is not locked for reading on `thread_id`.
+    fn read_unlock_account(&mut self, account: &Pubkey, thread_id: ThreadId) {
+        match self.read_locks.entry(*account) {
+            Entry::Occupied(mut entry) => {
+                let (thread_set, lock_counts) = entry.get_mut();
+                assert!(
+                    thread_set.contains(thread_id),
+                    "outstanding read lock must be on same thread"
+                );
+                lock_counts[thread_id] -= 1;
+                if lock_counts[thread_id] == 0 {
+                    thread_set.remove(thread_id);
+                    if thread_set.is_empty() {
+                        entry.remove();
+                    }
+                }
+            }
+            Entry::Vacant(_) => {
+                panic!("read lock must exist for account: {account}");
+            }
+        }
+    }
+}
+
+impl BitAndAssign for ThreadSet {
+    fn bitand_assign(&mut self, rhs: Self) {
+        self.0 &= rhs.0;
+    }
+}
+
+impl Sub for ThreadSet {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        Self(self.0 & !rhs.0)
+    }
+}
+
+#[cfg(test)]
+static_assertions::assert_eq_size!(ThreadSet, u64); // Formatting must change if size changes
+impl Display for ThreadSet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ThreadSet({:#064b})", self.0)
```

Comment:
length can be parameterized:

```patch
-#[cfg(test)]
-static_assertions::assert_eq_size!(ThreadSet, u64); // Formatting must change if size changes
 impl Display for ThreadSet {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "ThreadSet({:#064b})", self.0)
+        write!(f, "ThreadSet({:#0width$b})", self.0, width = MAX_THREADS)
     }
 }

```

Reply:
Wow, I had no idea that you could do that, that's awesome.

I guess one difference is, it looks like I can specify a runtime value for width - I wonder if the formatting is less efficient because it's not necessarily known at compile time?
Maybe it will optimize it since we are actually using a const?

Regardless of the efficiency, it's not intended for any performance critical paths and primarily meant as a debugging tool.

Reply:
yeah, it should be optimized well.

Reply:
done: fa3895a592ecad057d485c1facb814335f3520cd